agent/engine/task_manager_linux.go

```diff
@@ -15,30 +15,88 @@
 
 package engine
 
+import (
+	"github.com/aws/amazon-ecs-agent/agent/resources/cgroup"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+// control is used to manipulate cgroups and ease testing
+var control cgroup.Control
+
 // SetupPlatformResources sets up platform level resources
 func (mtask *managedTask) SetupPlatformResources() error {
-	if mtask.Task.CgroupEnabled() {
+	if mtask.engine.cfg.TaskCPUMemLimit {
+		setControl(cgroup.New())
 		return mtask.setupCgroup()
 	}
 	return nil
 }
 
 // CleanupPlatformResources cleans up platform level resources
 func (mtask *managedTask) CleanupPlatformResources() error {
-	if mtask.Task.CgroupEnabled() {
+	if mtask.engine.cfg.TaskCPUMemLimit {
 		return mtask.cleanupCgroup()
 	}
 	return nil
 }
 
 // setupCgroup sets up the cgroup for each managed task
 func (mtask *managedTask) setupCgroup() error {
-	// Stub for now
+	cgroupRoot, err := mtask.Task.BuildCgroupRoot()
+	if err != nil {
+		return errors.Wrap(err, "mtask setup cgroup: unable to determine cgroup root")
+	}
+
+	seelog.Debugf("Setting up cgroup at: %s", cgroupRoot)
+
+	if control.Exists(cgroupRoot) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", cgroupRoot)
+		return nil
+	}
+
+	linuxResourceSpec, err := mtask.Task.BuildLinuxResourceSpec()
+	if err != nil {
+		return errors.Wrapf(err, "mtask setup cgroup: unable to build resource spec")
+	}
+
+	cgroupSpec := cgroup.Spec{
+		Root:  cgroupRoot,
+		Specs: &linuxResourceSpec,
+	}
+
+	cgrp, err := control.Create(&cgroupSpec)
+	if err != nil {
+		return errors.Wrapf(err, "mtask setup cgroup: unable to create cgroup at %s", cgroupRoot)
+	}
+
+	// NOTE: This should be impossible
+	if cgrp == nil {
+		seelog.Criticalf("Invalid cgroup creation at %s", cgroupRoot)
+		return errors.New("mtask setup cgroup: invalid cgroup object")
+	}
+
 	return nil
 }
 
 // cleanupCgroup removes the task cgroup
 func (mtask *managedTask) cleanupCgroup() error {
-	// Stub for now
+	cgroupRoot, err := mtask.Task.BuildCgroupRoot()
+	if err != nil {
+		return errors.Wrap(err, "mtask cleanup cgroup: unable to determine cgroup root")
+	}
+
+	seelog.Debugf("Cleaning up cgroup at: %s", cgroupRoot)
+
+	err = control.Remove(cgroupRoot)
+	if err != nil {
+		return errors.Wrapf(err, "mtask cleanup cgroup: unable to remove cgroup at %s", cgroupRoot)
+	}
+
 	return nil
 }
+
+// setControl is used to set the cgroup controller
```

Comment:
Is this function just for testing?  If so, please add that to the comment (and you might also think about changing the name to `unsafeSetControl`).

Reply:
This is an anti-pattern.  I would like you to avoid doing this if possible.